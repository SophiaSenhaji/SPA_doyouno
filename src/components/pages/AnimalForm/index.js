import React, { Component } from 'react';
import { addOrUpdateAnimal, getAnimalById } from '../../../api/animal';
import Header from '../../Header';
import { getDate, getDateString } from '../../../utils/date';

const date = new Date();

const defaultState = {
    rfidNumber: '',
    species: 'Cat',
    race: '',
    height: '',
    weight: '',
    description: '',
    receptionDate: `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}`,
    birthDate: '',
    error: null
};
class AnimalForm extends Component {
    constructor() {
        super();
        const date = new Date();

        this.state = defaultState;
    }

    onInputChange = propertyName => event => {
        this.setState({
            [propertyName]: event.target.value
        })
    }

    validateForm = () => {
        const {
            rfidNumber,
            species,
            race,
            height,
            weight,
            description,
            receptionDate,
            birthDate
        } = this.state;
        const rfidRegex = new RegExp(/.{5}/)

        if (rfidNumber && !rfidRegex.test(rfidNumber)) return { message: 'Invalid rfidNumber (must be 5 characters)', property: 'rfidNumber' };
        if (!species) return { message: 'Invalid species', property: 'species' };
        if (isNaN(height) || !height || height < 0) return { message: 'Invalid height', property: 'height' };
        if (isNaN(weight) || !weight || weight < 0) return { message: 'Invalid weight', property: 'weight' };
        if (!description) return { message: 'Invalid description', property: 'description' };
        if (isNaN(getDate(receptionDate))) return { message: 'Invalid receptionDate', property: 'receptionDate' };
        if (birthDate && isNaN(getDate(birthDate))) return { message: 'Invalid birthDate', property: 'birthDate' };
        if (birthDate && birthDate > receptionDate) return { message: 'Birth date cannot be after reception date', property: 'receptionDate' };
    }

    onSubmit = async () => {   
        const { history } = this.props;    
        const {
            rfidNumber,
            species,
            race,
            height,
            weight,
            description,
            receptionDate,
            birthDate,
            id
        } = this.state;

        const animal = {
            id: this.state.id || undefined,
            rfidNumber: rfidNumber || undefined,
            species,
            race: race || undefined,
            height,
            weight,
            description,
            receptionDate,
            birthDate: birthDate || undefined
        }

        const error = this.validateForm();
        if (error) {
            this.setState({ error });
            return;
        }
        
        this.setState({ error: null });

        const result = await addOrUpdateAnimal(animal);
        if (result.message) {
            this.setState({ error: result });
            return;
        }

        history.push('/')

    }

    async componentDidMount() {
        const { match } = this.props;

        if (match.params.id) {
            const animal = await getAnimalById(match.params.id);
            this.setState({
                ...defaultState,
                ...animal,
                receptionDate: getDateString(animal.receptionDate),
                id: match.params.id
            });
        }

    }

    render() {
        const {
            rfidNumber,
            species,
            race,
            height,
            weight,
            description,
            receptionDate,
            birthDate,
            id,
            error
        } = this.state;
        return (
            <div className="AnimalForm">
                <Header withBack />
                <input placeholder="RFID number" type="text" onChange={this.onInputChange('rfidNumber')} value={rfidNumber || ''} />
                <select onChange={this.onInputChange('species')} value={species}>
                    <option>Cat</option>
                    <option>Dog</option>
                </select>
                <input placeholder="Race" type="text" onChange={this.onInputChange('race')} value={race || ''}/>
                <textarea placeholder="Description" onChange={this.onInputChange('description')} value={description}></textarea>
                <input placeholder="Weight" type="number" onChange={this.onInputChange('weight')} value={weight} />
                <input placeholder="Height" type="number" onChange={this.onInputChange('height')} value={height} />
                <input placeholder="Reception date (aaaa-mm-dd)" type="text" onChange={this.onInputChange('receptionDate')} value={receptionDate} />
                <input placeholder="Birth date (aaaa-mm-dd)" type="text" onChange={this.onInputChange('birthDate')} value={birthDate || ''} />
                {!id && <button onClick={this.onSubmit}>Add</button>}
                {id && <button onClick={this.onSubmit}>Edit</button>}
                {error && <div className="error">{error.message}</div>}
            </div>
        )
    }
}

export default AnimalForm;