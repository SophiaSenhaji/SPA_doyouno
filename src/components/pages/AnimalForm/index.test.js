import React from 'react';
import { shallow } from 'enzyme';
import AnimalForm from './index.js';
import toJson from 'enzyme-to-json';

describe('AnimalForm component', () => {
    const props = {
        match: {
            params: {}
        }
    };
    describe('validateForm', () => {
        it('should return an error if rfidNumber is invalid', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: 'ezd' });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Invalid rfidNumber (must be 5 characters)', property: 'rfidNumber' })
        })
    })
    describe('validateForm', () => {
        it('should return an error if species is invalid', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: '12345', species: undefined });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Invalid species', property: 'species' })
        })
        it('should return an error if height is not a number', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: '12345', species: 'Dog', height: 'height' });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Invalid height', property: 'height' })
        })
        it('should return an error if height is 0', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: '12345', species: 'Dog', height: 0 });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Invalid height', property: 'height' })
        })
        it('should return an error if height < 0', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: '12345', species: 'Dog', height: -1 });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Invalid height', property: 'height' })
        })
        it('should return an error if weight is not a number', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: '12345', species: 'Dog', height: 3, weight: 'weight' });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Invalid weight', property: 'weight' })
        })
        it('should return an error if weight is 0', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: '12345', species: 'Dog', height: 3, weight: 0 });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Invalid weight', property: 'weight' })
        })
        it('should return an error if weight < 0', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: '12345', species: 'Dog', height: 3, weight: -1 });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Invalid weight', property: 'weight' })
        })
        it('should return an error if description is invalid', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: '12345', species: 'Dog', height: 3, weight: 3, description: undefined });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Invalid description', property: 'description' })
        })
        it('should return an error if description is invalid', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: '12345', species: 'Dog', height: 3, weight: 3, description: 'description', receptionDate: 'receptionDate' });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Invalid receptionDate', property: 'receptionDate' })
        })
        it('should return an error if description is invalid', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: '12345', species: 'Dog', height: 3, weight: 3, description: 'description', receptionDate: '2019-02-05', birthDate: 'test' });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Invalid birthDate', property: 'birthDate' })
        })
        it('should return an error if rfidNumber is invalid', () => {
            const wrapper = shallow(<AnimalForm {...props} />);
            wrapper.setState({ rfidNumber: '12345', species: 'Dog', height: 3, weight: 3, description: 'description', receptionDate: '2019-02-05', birthDate: '2019-02-06' });
            const error = wrapper.instance().validateForm();
            expect(error).toEqual({ message: 'Birth date cannot be after reception date', property: 'receptionDate' })
        })
    })

    describe('render method', () => {
        it('should render correctly', () => {
            const fullProps = { ...props, rfidNumber: '12345', species: 'Dog', height: 3, weight: 3, description: 'description', receptionDate: '2019-02-06', birthDate: '2019-02-05' };
            const wrapper = shallow(<AnimalForm {...fullProps} />);
            expect(toJson(wrapper)).toMatchSnapshot();
        })
        it('should render correctly with id', () => {
            const fullProps = { ...props, rfidNumber: '12345', species: 'Dog', height: 3, weight: 3, description: 'description', receptionDate: '2019-02-06', birthDate: '2019-02-05' };
            const wrapper = shallow(<AnimalForm {...fullProps} id="id" />);
            expect(toJson(wrapper)).toMatchSnapshot();
        })
        it('should render correctly with error', () => {
            const fullProps = { ...props, rfidNumber: '12345', species: 'Dog', height: 3, weight: 3, description: 'description', receptionDate: '2019-02-06', birthDate: '2019-02-05' };
            const wrapper = shallow(<AnimalForm {...fullProps} error="error" />);
            expect(toJson(wrapper)).toMatchSnapshot();
        })
    })
})