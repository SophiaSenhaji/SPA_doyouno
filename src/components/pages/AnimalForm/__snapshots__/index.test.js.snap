// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimalForm component render method should render correctly 1`] = `
<div
  className="AnimalForm"
>
  <Header
    withBack={true}
  />
  <input
    onChange={[Function]}
    placeholder="RFID number"
    type="text"
    value=""
  />
  <select
    onChange={[Function]}
    value="Cat"
  >
    <option>
      Cat
    </option>
    <option>
      Dog
    </option>
  </select>
  <input
    onChange={[Function]}
    placeholder="Race"
    type="text"
    value=""
  />
  <textarea
    onChange={[Function]}
    placeholder="Description"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Weight"
    type="number"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Height"
    type="number"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Reception date (aaaa-mm-dd)"
    type="text"
    value="2019-2-5"
  />
  <input
    onChange={[Function]}
    placeholder="Birth date (aaaa-mm-dd)"
    type="text"
    value=""
  />
  <button
    onClick={[Function]}
  >
    Add
  </button>
</div>
`;

exports[`AnimalForm component render method should render correctly with error 1`] = `
<div
  className="AnimalForm"
>
  <Header
    withBack={true}
  />
  <input
    onChange={[Function]}
    placeholder="RFID number"
    type="text"
    value=""
  />
  <select
    onChange={[Function]}
    value="Cat"
  >
    <option>
      Cat
    </option>
    <option>
      Dog
    </option>
  </select>
  <input
    onChange={[Function]}
    placeholder="Race"
    type="text"
    value=""
  />
  <textarea
    onChange={[Function]}
    placeholder="Description"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Weight"
    type="number"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Height"
    type="number"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Reception date (aaaa-mm-dd)"
    type="text"
    value="2019-2-5"
  />
  <input
    onChange={[Function]}
    placeholder="Birth date (aaaa-mm-dd)"
    type="text"
    value=""
  />
  <button
    onClick={[Function]}
  >
    Add
  </button>
</div>
`;

exports[`AnimalForm component render method should render correctly with id 1`] = `
<div
  className="AnimalForm"
>
  <Header
    withBack={true}
  />
  <input
    onChange={[Function]}
    placeholder="RFID number"
    type="text"
    value=""
  />
  <select
    onChange={[Function]}
    value="Cat"
  >
    <option>
      Cat
    </option>
    <option>
      Dog
    </option>
  </select>
  <input
    onChange={[Function]}
    placeholder="Race"
    type="text"
    value=""
  />
  <textarea
    onChange={[Function]}
    placeholder="Description"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Weight"
    type="number"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Height"
    type="number"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Reception date (aaaa-mm-dd)"
    type="text"
    value="2019-2-5"
  />
  <input
    onChange={[Function]}
    placeholder="Birth date (aaaa-mm-dd)"
    type="text"
    value=""
  />
  <button
    onClick={[Function]}
  >
    Add
  </button>
</div>
`;
